### ISSUE TITLE
[BUG]: Owned keyword not working as expected with Immutable Int

### ISSUE THREAD

**Issue by Tucker6742** [modularml/mojo#747]
### Bug description

When using owned with immutable Int in a function, the variable is destroy after running the function
According to the [documentation](https://docs.modular.com/mojo/manual/basics/#function-arguments-and-returns) the variable inside the function should just be a copy and not affect the variable outside

> Another option is to declare the argument as owned, which provides the function full ownership of the value (it’s mutable and guaranteed unique). **<ins>This way, the function can modify the value and not worry about affecting variables outside the function.</ins>**

**testing.mojo**
```
fn owned_int(owned x: Int) -> Int:
    x += 1
    return x


fn main():
    let a = 1
    print(a)
    let b = owned_int(a)
    print(b)
    print(a)
```
Running the code give this error
```
testing.mojo:11:10: error: use of uninitialized value 'a'
    print(a)
         ^
testing.mojo:7:5: note: 'a' declared here
    let a = 1
    ^
~/.modular/pkg/packages.modular.com_mojo/bin/mojo: error: failed to run the pass manager
```

The issue doesn't appear when using owned with a string

```
fn owned_string(owned x: String) -> String:
    x += "_1"
    return x


fn main():
    let a = "abc"
    print(a)
    let b = owned_string(a)
    print(b)
    print(a)
```
Compiler result
```
abc
abc_1
abc
```

In the borrowed_int example if we assign the ```a``` with key word ```var``` then the error will not occur but there will be a warning that ```a``` is declare with ```var``` but not mutated and should just be declare with ```let```

```
fn owned_int_with_var(owned x: Int) -> Int:
    x += 1
    return x


fn main():
    var a = 1
    print(a)
    let b = owned_int_with_var(a)
    print(b)
    print(a)
```
Compiler result
```
testing.mojo:7:5: warning: 'a' was declared as a 'var' but never mutated, consider switching to a 'let'
    var a = 1
    ^
1
2
1
```

### Steps to reproduce

Running each of the code above

### System information

```shell
- What OS did you do install Mojo on ?: Windows 10 WSL Ubuntu 20.04
- Provide version information for Mojo by pasting the output of `mojo -v`: mojo 0.2.1 (64d14e85)
- Provide Modular CLI version by pasting the output of `modular -v`: modular 0.1.4 (6b54d308)
```

**Comment by Mogball** [comment]
This is weird. Thanks for filing! I'll take a look

**Comment by chenweichen** [comment]
I'm encountering the same problem.
I'm using modular playground.
The version of mojo kernel should be v0.4.0 at the time.  

According to [Argument mutability and ownership](https://docs.modular.com/mojo/manual/basics/#argument-mutability-and-ownership) 
I modified the function from the example in the docs.    

My modified example:
```
fn add_owned(owned x: Int, owned y: Int) -> Int:
    x += 10
    y += 20
    return x + y
```

### Situation 1:
a and b are declared as **var**

```
fn main_1():
    var a: Int = 100
    var b: Int = 200
    print("a: ", a, "b: ", b)
    let c = add_owned(a, b)
    print("c: ", c)
    print("a: ", a, "b: ", b)
    
main_1()
```

The result:
```
a:  100 b:  200
c:  330
a:  100 b:  200
```  
It works.  


### Situation 2
a and b are declared as **let**
```
fn main_2():
    let a: Int = 100
    let b: Int = 200
    print("a: ", a, "b: ", b)
    let c = add_owned(a, b)
    print("c: ", c)
    print("a: ", a, "b: ", b)
    
main_2()
```

The result:
```
expression failed to parse (no further compiler diagnostics)
```

It looks like Mojo didn't make copies of a and b.

**Comment by b4rdos** [comment]
Adding my voice to this. I have came across the same issue. Started a conversation in [discord](https://discord.com/channels/1087530497313357884/1166282705873416222).

TL;DR

Following the docs for [String](https://docs.modular.com/mojo/programming-manual.html#transfer-arguments-owned-and) and [UniquePointer](https://docs.modular.com/mojo/programming-manual.html#transfer-arguments-owned-and) produced expected results. Modifying the example to use `Int`, `Float32`, and `Bool` produced and error along the lines of 
```
/path/to/hello-world/take_ownership.mojo:9:26: error: use of uninitialized value 'x'
```
This usually points to the line with a function call after the variable has been passed to one with `owned` in the argument.

I suspect it relates to stack vs heap memory.

A related issue is the fact the VSCode extension did not pick up on this issue when writing the code. I only encountered this during `build`.

**Comment by ilkkanisula** [comment]
I noticed this also and compiled output is identical with or without owned.

I created short shell script to compile owned and not_owned functions and dissassemble binaries and compare them.
See here: [test_owned.sh](https://gist.github.com/ilkkanisula/5561135ed4cfa92c72d4f3609f020eca).

# Reproduce
Preconditions: Mojo sdk must be installed alreay
Run: Copy script to some empty directory and run ./test_owned.sh
Results: Disassemblies are identical (should be different?)

Script creates src and output directories and two .mojo code files compiles, dissassembles and compares outputs.

```
% ./test_owned.sh || echo error
Creating src and out directories
Build src/owned.🔥 to out/owned
Disassemble out/owned to out/owned.objdump
Build src/not_owned.🔥 to out/not_owned
Disassemble out/not_owned to out/not_owned.objdump
Compare dissaaemblies of owned and not_owned: Identical
error
```

**Comment by Mogball** [comment]
@stumpOS when you have some free cycles, can you take a look?

**Comment by lattner** [comment]
Thank you for filing this, this is fixed in the next update!
